### PR TITLE
#332 postcode extraction fix

### PR DESCRIPTION
**Pull request by denironyx**
This PR also updates postcode extraction so we avoid running the broad UK postcode regex over `address_concat` when a postcode has already been supplied.

Previously, postcode like address fragments could be incorrectly stripped before normalisation. For example:

- `FLAT D1 2ND FLR EXAMPLE HOUSE LONDON`
- `UNIT Q3 2ND FLR EXAMPLE WORKS LONDON`

The regex interpreted `D1 2ND` / `Q3 2ND` as postcode-like substrings, removed them, and prevented `2ND FLR` from normalising to `SECOND FLOOR`.

The new behaviour is:

- if `postcode` is supplied, keep `address_concat` unchanged and use the supplied postcode
- if `postcode` is missing, retain the existing regex based extraction/removal from `address_concat`

This keeps the fix simple and avoids adding extra regex work on the common path where postcode is already present. The tradeoff is that a duplicated postcode may remain in `clean_full_address` when `postcode` is supplied, but that is preferable to removing valid address tokens.

Added regression coverage for the `D1 2ND` / `Q3 2ND` cases, plus a missing postcode case to confirm postcode extraction still works.

Workflow output:

```
┌───────────┬──────────────────────────────────────┬──────────┬────────────┐
│ unique_id │            address_concat            │ postcode │ ukam_label │
│  varchar  │               varchar                │ varchar  │  varchar   │
├───────────┼──────────────────────────────────────┼──────────┼────────────┤
│ a         │ FLAT D1 2ND FLR EXAMPLE HOUSE LONDON │ SE1 2AB  │ 1          │
│ b         │ UNIT Q3 2ND FLR EXAMPLE WORKS LONDON │ SE1 2AB  │ 2          │
│ c         │ FLAT A 3RD FLR EXAMPLE HOUSE LONDON  │ SE1 2AB  │ 3          │
└───────────┴──────────────────────────────────────┴──────────┴────────────┘

┌───────────┬──────────────────────────────────────┬──────────┬─────────────────────────┬────────────────────────────────────────────┬─────────────────┬─────────────────┬────────────────┬─────────┬────────────┬─────────┬───────────┐
│ unique_id │       original_address_concat        │ postcode │ postcode_like_substring │             clean_full_address             │ flat_positional │  flat_identity  │ numeric_tokens │ has_2nd │ has_second │ has_3rd │ has_third │
│  varchar  │               varchar                │ varchar  │         varchar         │                  varchar                   │     varchar     │     varchar     │   varchar[]    │ boolean │  boolean   │ boolean │  boolean  │
├───────────┼──────────────────────────────────────┼──────────┼─────────────────────────┼────────────────────────────────────────────┼─────────────────┼─────────────────┼────────────────┼─────────┼────────────┼─────────┼───────────┤
│ a         │ FLAT D1 2ND FLR EXAMPLE HOUSE LONDON │ SE1 2AB  │ D1 2ND                  │ FLAT D 1 SECOND FLOOR EXAMPLE HOUSE LONDON │ SECOND FLOOR    │ _D_SECOND FLOOR │ [1]            │ false   │ true       │ false   │ false     │
│ b         │ UNIT Q3 2ND FLR EXAMPLE WORKS LONDON │ SE1 2AB  │ Q3 2ND                  │ UNIT Q 3 SECOND FLOOR EXAMPLE WORKS LONDON │ SECOND FLOOR    │ __SECOND FLOOR  │ [3]            │ false   │ true       │ false   │ false     │
│ c         │ FLAT A 3RD FLR EXAMPLE HOUSE LONDON  │ SE1 2AB  │                         │ FLAT A THIRD FLOOR EXAMPLE HOUSE LONDON    │ THIRD FLOOR     │ _A_THIRD FLOOR  │ []             │ false   │ false      │ false   │ true      │
└───────────┴──────────────────────────────────────┴──────────┴─────────────────────────┴────────────────────────────────────────────┴─────────────────┴─────────────────┴────────────────┴─────────┴────────────┴─────────┴───────────┘

```

### CODE REVIEW

**Comment by denironyx** [comment]
@ThomasHepworth what do you think about this?

File: uk_address_matcher/cleaning/steps/normalisation.py [hunk 1]
```diff
@@ -55,27 +55,47 @@ def _extract_postcode_from_address() -> str:
...
     uk_postcode_regex = r"\b(?:GIR ?0AA|[A-Z][A-HJ-Y]?\d[A-Z\d]? ?\d[A-Z]{2})\b"
 
-    # Extract postcode from address and remove it from address_concat
-
-    # - Use COALESCE to prefer existing postcode over extracted
-    # - Always remove postcode from  address_concat to avoid duplication
+    # Extract postcode from address only when one has not been supplied.
+    # When a postcode is supplied, strip postcode-like suffixes only. Running
+    # the broad regex over the whole address can remove valid fragments such as
+    # "D1 2ND" before floor normalisation runs.
```
Comment (ThomasHepworth) [inline uk_address_matcher/cleaning/steps/normalisation.py:61]: Ah, I was under the impression this code only ran in situations in which we were missing a postcode column. Very happy to approve this.

File: tests/cleaning/test_cleaning_steps.py [hunk 2]
```diff
@@ -266,6 +267,60 @@ def test_remove_duplicate_end_tokens():
...
+        ),
+        (
+            "c",
+            "SE1 2AB",
+            "FLAT A THIRD FLOOR EXAMPLE HOUSE LONDON",
+        ),
+        ("d", "SW1A 2AA", "10 DOWNING STREET"),
+        ("e", "SW1A 2AA", "10 DOWNING STREET"),
+    ]
+
```
Comment (ThomasHepworth) [inline tests/cleaning/test_cleaning_steps.py:322]: What happens in situations where an invalid postcode is provided?
```
postcode: "N/A"
address_concat: "10 Downing Street, London SW1A 2AA"
```

File: uk_address_matcher/cleaning/steps/normalisation.py (hunk 1 above)
Comment (denironyx) [reply uk_address_matcher/cleaning/steps/normalisation.py:61]: I think these are edge cases, that needed to be accounted for as well.

File: tests/cleaning/test_cleaning_steps.py (hunk 2 above)
Comment (denironyx) [reply tests/cleaning/test_cleaning_steps.py:322]: It takes the postcode from the `addess_concat` string. In this case `SW1A 2AA`.

**Review by ThomasHepworth — APPROVED** [review]
Sorry for the delay - happy to get this merged in following your replies!
